auditbeat/helper/hasher/hasher.go

```diff
@@ -0,0 +1,285 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package hasher
+
+import (
+	"crypto/md5"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/hex"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"time"
+
+	"github.com/OneOfOne/xxhash"
+	"github.com/dustin/go-humanize"
+	"github.com/joeshaw/multierror"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/blake2b"
+	"golang.org/x/crypto/sha3"
+	"golang.org/x/time/rate"
+
+	"github.com/elastic/beats/libbeat/common/file"
+)
+
+// HashType identifies a cryptographic algorithm.
+type HashType string
+
+// Unpack unpacks a string to a HashType for config parsing.
+func (t *HashType) Unpack(v string) error {
+	*t = HashType(v)
+	return nil
+}
+
+// IsValid checks if the hash type is valid.
+func (t *HashType) IsValid() bool {
+	_, valid := validHashes[*t]
+	return valid
+}
+
+var validHashes = map[HashType]struct{}{
+	BLAKE2B_256: {},
+	BLAKE2B_384: {},
+	BLAKE2B_512: {},
+	MD5:         {},
+	SHA1:        {},
+	SHA224:      {},
+	SHA256:      {},
+	SHA384:      {},
+	SHA512:      {},
+	SHA512_224:  {},
+	SHA512_256:  {},
+	SHA3_224:    {},
+	SHA3_256:    {},
+	SHA3_384:    {},
+	SHA3_512:    {},
+	XXH64:       {},
+}
```

Comment:
You can make this map store the constructors: `map[HashType]func()hash.Hash`. 

This helps avoid the big switch statement in HashFile below. Also makes defining the separate HashType constants less necessary, so all can be kept in just one place.

Reply:
I think this would work if the constructors of all hashes would conform to `func() hash.Hash`. Unfortunately, not all of them do, e.g. blake2b is `func New256(key []byte) (hash.Hash, error)`.

Reply:
In that case you can do something like:
```
BLAKE2B_256: func() hash.Hash {
    h,_ := blake2b.New256(nil)
    return h
}
```

But again, this may be out of scope for this PR.

Reply:
I like it, let me do that. Thanks!